### DISCUSSION TITLE
UR10 Remote Connection issues (network problem?)

### DISCUSSION THREAD

**Discussion by UnfoldAntwerp** [visose/Robots#156] [Q&A] [answered]
Hi all,

I'm running (again) into stability issues with the remote connection to a UR 10 now that I moved everything to a different location and network setup. Sending is very unreliable, sometimes it comes through, often not. Pressing the send button longer or more frequently seems to solve it once in a while but especially when the amount of targets gets higher, it fails. We tried multiple computers, some experience the problem more than others but it's on every computer unreliable...

A bit of backstory:
When I first started testing the grasshopper with UR10 at the university where it was located, I ran into similar issues. After a bit more testing and playing around they disappeared but we don't know why. One suspicion was that we were sending too many targets, something that was hinted at in [this discussion](https://github.com/visose/Robots/discussions/112) but once our setup was stable we tested increasingly higher target counts and got to 50.000 without issues, it just took long to send with an reliable rate of 370 targets per second (see that post for my upload countdown counter). So I was under the impression that the problem was solved and one just had to be patient when sending bigger files.

But now we moved the setup to an exhibition space with a dedicated wireless router and the issues come up again, sending is highly unreliable when going over a few 100 targets. The robot is connected to the ethernet with fixed IP and computers were connecting to the wireless. The router is not attached to internet but is handling DHCP so each computer gets an IP. When I switched the sole desktop computer to wired ethernet computer through the router I managed to get a more reliable connection and succesfully managed to send 18k targets but still this was with a bit hit n miss, and it seemed to come through faster than my usual 370 targets/s. The Mac computers where also more successful than the PC laptop so far.

So now I'm thinking it has something to do with the network connection to the robot. But I'm a bit in the dark as how I can test what the issue is and how to solve it.

-is there a method for Mac or PC to verify or check for network issues, latency, dropped connections etc?
-is there a direct way to connect to the ethernet connection on UR10 without a router? Can the robot setup a network?
alternatively, how does one create a program from the script you can save from Robots to load locally on the pendant if network connection keeps failing us?
-anyone seeing these issues too? @robin-gdwl, you mentioned this in the other post

All help & hints tremendously appreciated!

**Comment by visose** [comment]
You can connect the UR controller directly to the pc with a network cable, no need for a router. You have to set a fixed IP on both your computer and the UR controller. You can check for package issues opening a terminal (`cmd`, `powershell` in Windows, `Terminal` in macOS) and type `ping robotIPAddress`. This will show if you are reaching the robot and if any packet loss.

In Grasshopper, use the `save program` component to save the UR script file. Then either copy it to a USB stick or use the ftp server in the controller. For this option, install [FileZilla](https://filezilla-project.org/download.php) (since it needs to support SFTP). Use the following:
Hotname: `sftp://robotIPAddress`
User: `root`
password: `easybot`
Port: `22`

After copying the file, you can create an empty program in the controller using the pendant and load the script there.

There will be a size limit to the program due to what is mentioned in the other discussion, but it's likely you are not reaching that limit and it's a different issue.

**Reply by UnfoldAntwerp** [reply]
Thanks! I'll check tomorrow and report back!

**Reply by UnfoldAntwerp** [reply]
@visose, did a whole day of testing but unfortunately the direct connection did not make a meaningful change over the connection over a router. The odd thing is that we have one Mac laptop that consistently can send high target count files to the robot over wireless through the router. Another mac is partially reliable over ethernet through router, same with direct connection, but less so when wireless. The PC is unable to send anything more than couple of targets, doesn't matter if that is over direct connection, wired through router or wireless... We did several ping tests. No meaningful difference between direct and over router when using cable. All below 1ms and no dropped packets when sending 500 pings. When wireless, the average is around 16ms but with regular highs of ±250ms. But i'm afraid now that network connection is not the issue since one computer connects reliably over wifi and it was running fine on a complex university network. And pinging seems to be reliable. I'm totally in the dark why there is this flux in reliability. Would Robots version matter? The reliable machine runs v0.0.5, the semi reliable too and the unreliable 1.1.3.

In regards to the file upload over FTP or USB. I can't get it to run when I load the script in a new program. It just refuses to run... I need to look deeper into that.

**Reply by visose** [reply]
I don't think anything has changed regarding UR network. There were some changes to the UR code generation to make it a bit cleaner, but this shouldn't have any effect. There should be other users using UR robots but haven't heard of any similar issues.

Is the Mac laptop running Windows? Could you update it to the latest release and see if it's still reliable?

If you can figure out the conditions that lead to reproducing the issue that would be great, since that means I could reproduce it here and fix the issue. For example, when you say unreliable, does this mean that the exact same program sometimes works or sometimes not, or is it different programs (some programs always work, some always don't). Is it the length of the program, or is it the use of different motion types, commands, etc.).

**Reply by UnfoldAntwerp** [reply]
@visose I'll try to do some more digging and write a better description. I'm usually good at tracing a problem down to something reproducible but here I'm a bit in the dark as it is pretty random. But short answers before I can do more testing which can take some time:
-The Mac laptop is running macOS. I'll try to update to latest Robots once this exhibition is over as I don't want to fiddle with the only working setup.
-Unreliable means a single GH definition without any changes to the program will sometimes send and sometimes not. Within this GH definition I can switch geometries for printing and with increasingly larger or more complex geometries the target count goes up. Now with this increasing target count, the rate of a program arriving at the UR10 successfully declines with the amount of targets I'm trying to send. So for example (this is out of my head, not recorded nrs) I'm being able to send a 200 target program successfully 8/10 times when I press the send button, a 1000 target program 5/10 times, a 3000 target program 2/10 times and a 10.000 targets program 0/10 times. On another computer the same GH definition is able to send 5/10 times for 200 targets, 1/10 times for 1000 targets and never above. The last computer is able to send anything (at least up to 20k tested) without a hitch...  Would be great if I had more diagnostic powers here.

ps. I was able to load programs by USB/SFTP and run them on the UR10 successfully now using the described method but I had to delete the first line `def Program():`  from the saved script and change the extension to .script or select 'all files' on the UR controller.

**Reply by robin-gdwl** [reply]
Note: In my testing with URsim and FTP connections the username set by default is not "root" but "ur". 
I can not get FTP connections with root:easybot on a clean install of the latest version of URsim. 
ur:easybot works.

**Comment by visose** [comment]
There is now support to upload the program via FTP using the Remote component, this might fix the issue when streaming large programs. More info here: https://github.com/visose/Robots/discussions/107#discussioncomment-2775301

**Comment by visose** [comment, marked as answer]
Closing this as likely resolved by the FTP/SFTP upload path added later. For large UR programs, use the Remote component with an `sftp://...` IP/connection string, or use Save Program and transfer the generated program. If secondary-client streaming is still unreliable on a current version, please reopen with the exact Robots version, controller version, network path, and a minimal reproducible file.
